src/workers/helpers.js

```diff
@@ -1,53 +1,238 @@
+}
+
+export function postSuccess(step, info) {
+  if (typeof info == "undefined") {
     postMessage({
-        type: `${step}_START`
+      type: `${step}_CACHE`,
     });
+  } else {
+    var transferable = [];
+    extractBuffers(info, transferable);
+    postMessage(
+      {
+        type: `${step}_DATA`,
+        resp: info,
+      },
+      transferable
+    );
+  }
 }
 
-export function postSuccess(step, info) {
-    if (typeof info == "undefined") {
-        postMessage({
-            type: `${step}_CACHE`
-        });
-    } else {
-        var transferable = [];
-        extractBuffers(info, transferable);
-        postMessage({
-            type: `${step}_DATA`,
-            resp: info
-        }, transferable);
+export function postError(type, err, fatal) {
+  postMessage({
+    type: `${type}_ERROR`,
+    resp: {
+      reason: err.toString(),
+      fatal: fatal,
+    },
+  });
+}
+
+export function splitMetricsByBlock(metrics, blockLevels, blockIds) {
+  var output = {};
+  var blocks = blockIds.array();
+  for (var b = 0; b < blockLevels.length; b++) {
+    let current = {};
+    for (const [key, val] of Object.entries(metrics)) {
+      current[key] = val.array().filter((x, i) => blocks[i] == b);
     }
+    output[blockLevels[b]] = current;
+  }
+  return output;
 }
 
-export function postError(type, err, fatal) {
-    postMessage({
-        type: `${type}_ERROR`,
-        resp: {
-            reason: err.toString(),
-            fatal: fatal
-        },
-    });
+export function splitThresholdsByBlock(thresholds, blockLevels) {
+  var output = {};
+  for (const x of blockLevels) {
+    output[x] = {};
+  }
+
+  for (const [key, val] of Object.entries(thresholds)) {
+    for (var b = 0; b < blockLevels.length; b++) {
+      output[blockLevels[b]][key] = val[b];
+    }
+  }
+
+  return output;
+}
+
+export async function fetchStepSummary(state, step) {
+  // do not send any response to UI if they have not changed
+  if (!state[step].changed) {
+    return null;
+  }
+
+  if (step === "inputs") {
+    let output = {};
+
+    let ngenes = {};
+    for (const a of state[step].fetchCountMatrix().available()) {
+      ngenes[a] = state[step].fetchCountMatrix().get(a).numberOfRows();
+    }
+
+    let gene_info = {};
+    for (const [k, v] of Object.entries(
+      state[step].fetchFeatureAnnotations()
+    )) {
+      let info = {};
+      for (const c of v.columnNames()) {
+        info[c] = bioc.CLONE(v.column(c));
+      }
+      gene_info[k] = info;
+    }
+
+    output = {
+      num_cells: state[step].fetchCountMatrix().numberOfColumns(),
+      num_genes: ngenes,
+      genes: gene_info,
+      annotations: state[step].fetchCellAnnotations().columnNames(),
+    };
+
+    return output;
+  } else if (step === "rna_quality_control") {
+    let output = {};
+    var blocks = state["inputs"].fetchBlockLevels();
+    if (blocks === null) {
+      blocks = ["default"];
+      output.data = { default: state[step].fetchMetrics() };
```

Comment:
`fetchMetrics()` now returns a reference to a Wasm object, this needs to be stripped out with

https://github.com/kanaverse/bakana/blob/c0d746d6aa7893da81daeecf7404694e97e814eb/src/steps/quality_control.js#L198-L204